src/main/java/com/vinyl/service/UserService.java

```diff
@@ -59,10 +64,23 @@ public void deleteUser(EmailPassDTO credentials) {
 		user.ifPresent(userRepository::delete);
 	}
 
-	public String loginUser(EmailPassDTO loginInfo) {
+	public Token loginUser(EmailPassDTO info) {
 
-		validatorFactory.getEmailAndPasswordValidator().validate(loginInfo);
-		return "dummy token";
+		validatorFactory.getEmailAndPasswordValidator().validate(info);
+
+		Token token = new Token();
+		
+		token.setUser(userRepository.findByEmail(info.getEmail()).get());
+		token.setHash(String.valueOf(Math.abs(info.hashCode())));
+
+		return save(token);
+	}
 
+	public Token save(Token toSave) {
+		Token token = tokenRepository.findByHash(toSave.getHash());
+		if (token == null) {
```

Comment:
if token is null you try to save the null in the db?

Reply:
if token is not in the database, then i save the toSave token given by the param.
so if token is already in the database ( token != null ) i just return it. 
the token variable is actually the stored token in the database.

Reply:
ok, then it is a little wierd, you do a lot of work (creating the token, populating it's fields, etc) and then you just discard that work if a token already exists in the db.
You should check if a token exists at the beginning of the login method, then check if it is still valid (current time is not greater than validUntil) if it is return in, otherwise create a new one and return it